File: benchmark_read_NIfTI.R
if (!require("oro.nifti", quietly = TRUE))
{
  install.packages("oro.nifti")
}
library(oro.nifti)

if (!require("RNifti", quietly = TRUE))
{
  install.packages("RNifti")
}
library(RNifti)

if (!require("this.path", quietly = TRUE))
{
  install.packages("this.path")
}
library(this.path)

if (!require("BiocManager", quietly = TRUE))
{
 install.packages("BiocManager")
}

if (!require("rhdf5", quietly = TRUE))
{
 BiocManager::install("rhdf5")
}
library(rhdf5)

# Get the path of this script
workDir    <- this.path::this.dir()
inDir      <- file.path(workDir, "samples", "NIfTI")
resultsDir <- file.path(workDir, "results")
dir.create(resultsDir, showWarnings = FALSE, recursive = TRUE)

# Paths to concatenated data
file_compressed   <- file.path(inDir, "concatedData_73_compressed.mat")
file_uncompressed <- file.path(inDir, "concatedData_73_uncompressed.mat")

# Make a list of NIfTI files
listNIfTI <- list.files(inDir, pattern = glob2rx("*.nii.*"))

# Initialize
numRepeats            <- 10
t_readNIfTI_oro       <- vector("numeric", length = numRepeats)
t_readNIfTI_rnifti    <- vector("numeric", length = numRepeats)
t_rhdf5_compressed    <- vector("numeric", length = numRepeats)
t_rhdf5_uncompressed  <- vector("numeric", length = numRepeats)

for (rep in 1:numRepeats)
{
  # Read NIfTI images
  t_readNIfTI_oro[rep]      <- system.time(for (file in listNIfTI){nii <- readNIfTI(file.path(inDir, file))})["elapsed"]
  t_readNIfTI_rnifti[rep]   <- system.time(for (file in listNIfTI){nii <- readNifti(file.path(inDir, file))})["elapsed"]
  t_rhdf5_compressed[rep]   <- system.time(h5read(file_compressed,   name = "/"))["elapsed"]
  t_rhdf5_uncompressed[rep] <- system.time(h5read(file_uncompressed, name = "/"))["elapsed"]
}

# Show results
cat(paste0("Time taken: readNIfTI (oro): ",       round(median(t_readNIfTI_oro), 4), "\n"))
cat(paste0("Time taken: readNifti (RNifti): ",    round(median(t_readNIfTI_rnifti), 4), "\n"))
cat(paste0("Time taken: h5read (compressed): ",   round(median(t_rhdf5_compressed), 4), "\n"))
cat(paste0("Time taken: h5read (uncompressed): ", round(median(t_rhdf5_uncompressed), 4), "\n"))


# Clear up and save
rm(nii)
save.image(file=file.path(resultsDir, "benchmarks_readNIfTI_R.rdata"))

File: benchmark_read_NIfTI.m
function benchmark_read_NIfTI(inDir, resultsDir)
%% Code to benchmark reading NIfTI files and concatenated data
% Set paths, relative to this script
if ~exist('resultsDir', 'var') || isempty(resultsDir)
    workDir     = fileparts(mfilename('fullpath'));
    resultsDir  = fullfile(workDir, 'results');
end

if ~exist('inDir', 'var') || isempty(inDir)
    workDir = fileparts(mfilename('fullpath'));
    inDir   = fullfile(workDir, 'samples', 'NIfTI');
end

if ~exist(resultsDir, 'dir')
    mkdir(resultsDir);
end

% Define functions
fUncompressed   = @() load(fullfile(inDir, 'concatedData_73_uncompressed.mat'), 'data', 'vec_mask');
fCompressed     = @() load(fullfile(inDir, 'concatedData_73_compressed.mat'), 'data', 'vec_mask');
fUncompressedH5 = @() h5read_mat(fullfile(inDir, 'concatedData_73_uncompressed.mat'));
fCompressedH5   = @() h5read_mat(fullfile(inDir, 'concatedData_73_compressed.mat'));
fNIfTI          = @() read_NIfTI(inDir);

% Get robust timing using timeit
tRead_uncompressed   = timeit(fUncompressed);
tRead_compressed     = timeit(fCompressed);
tRead_uncompressedH5 = timeit(fUncompressedH5);
tRead_compressedH5   = timeit(fCompressedH5);
tRead_NIfTI          = timeit(fNIfTI);

% Show results
disp(['Time taken: read NIfTI: ',                    num2str(tRead_NIfTI)]);
disp(['Time taken: read compressed HDF5 (load): ',   num2str(tRead_compressed)]);
disp(['Time taken: read uncompressed HDF5 (load): ', num2str(tRead_uncompressed)]);
disp(['Time taken: read compressed HDF5 (h5): ',     num2str(tRead_compressedH5)]);
disp(['Time taken: read uncompressed HDF5 (h5): ',   num2str(tRead_uncompressedH5)]);

% Save results
clear f*
save(fullfile(resultsDir, 'benchmarks_readNIfTI.mat'));
end

function read_NIfTI(inDir)
listFiles = dir(fullfile(inDir, '*.nii.gz'));

% Get dimensions
info = niftiinfo(fullfile(listFiles(1).folder, listFiles(1).name));

% Initialize
data = zeros([info.ImageSize, length(listFiles)], 'single');

% Loop over every subject and concat
for subjs = 1:length(listFiles)
    data(:, :, :, subjs) = niftiread(fullfile(listFiles(subjs).folder, listFiles(subjs).name));
end
end

function res = h5read_mat(inFile)
% See also, https://www.mathworks.com/matlabcentral/answers/329759
info = h5info(inFile);
toLoad = strcat({'/'}, {info.Datasets(:).Name});

% Initialize
res = cell(length(toLoad),1);

% Keep dataset open
fid = H5F.open(inFile);

for ii = 1:length(toLoad)
    try
        did = H5D.open(fid, toLoad{ii});
        res{ii} = H5D.read(did);
        H5D.close(did);
    catch
        H5F.close(fid);
    end
end
H5F.close(fid);
end